File: exercises/exercises-mpi-p/cubegather.py
#!/usr/bin/env python
################################################################
################################################################
####
#### This program file is part of the book and course
####   Parallel Computing for Science and Engineering
#### by Victor Eijkhout, copyright 2013-2020
####
#### cubegatherp.py : use subarray type
####
################################################################
################################################################

from mpi4py import MPI
import math
import numpy as np
import random
import sys

comm = MPI.COMM_WORLD

nprocs = comm.Get_size()
procno = comm.Get_rank()

##
## See if you can arrange the processes
## in a perfect cube. If not, quit.
##
procs_per_side = int( math.pow(nprocs+1,1./3) )
if pow(procs_per_side,3) != nprocs:
    if procno==0:
        print("Number of processors needs to be a perfect square")
    comm.Barrier()
    comm.Abort(0)

##
## Every process gets its own data and sends it to process zero.
## Exercise: why are we using Isend here?

mydata = np.zeros( 1,dtype=np.intc )
mydata[0] = procno+1
send_request = comm.Isend( mydata,0 )

if procno==0:
    cubedata = np.zeros(nprocs,dtype=np.intc)
    sizes    = np.array( [procs_per_side,procs_per_side,procs_per_side] )
    subsizes = np.array( [1,1,1] )
    for proci in range(procs_per_side):
        for procj in range(procs_per_side):
            for prock in range(procs_per_side):
                ##
                ## Set up a subarray type describing there the data
                ## from process (proci,procj,prock) goes in the cube
                ##
#### your code here ####
                insert_type.Commit()
                sender = ( proci*procs_per_side + procj )*procs_per_side + prock
                comm.Recv( [cubedata,1,insert_type],sender )
    print("Cube received: {}".format(cubedata))
    error = 0
    for p in range(nprocs):
        if cubedata[p]!=p+1:
            error += 1
    if error==0:
        print("Finished. Cube correctly gathered")
    else:
        print("There were errors")

##
## Conclude the sends
##
MPI.Request.Wait( send_request )
                

File: exercises/exercises-mpi-mpl/cubegather.cxx
/****************************************************************
 ****
 **** This program file is part of the book 
 **** `Parallel programming for Science and Engineering'
 **** by Victor Eijkhout, eijkhout@tacc.utexas.edu
 ****
 **** copyright Victor Eijkhout 2012-2020
 ****
 **** MPI Exercise for the subarray type, MPL version
 ****
 ****************************************************************/

#include <iostream>
#include <iomanip>
#include <sstream>
#include <cmath>
#include <random>
using namespace std;
#include <mpl/mpl.hpp>

int main() {

  const mpl::communicator &comm_world = mpl::environment::comm_world();
  int nprocs, procno;
  
  // compute communicator rank and size
  nprocs = comm_world.size();
  procno = comm_world.rank();

  stringstream proctext;
  
  /*
   * See if you can arrange the processes
   * in a perfect cube. If not, quit.
   */
  int procs_per_side;
  for (procs_per_side=1; ; procs_per_side++) {
    if (procs_per_side*procs_per_side*procs_per_side==nprocs)
      break;
    if (procs_per_side*procs_per_side*procs_per_side>nprocs) {
      if (procno==0) {
	proctext << "Number of processes needs to be a cube" << endl;
	cerr << proctext.str(); proctext.clear();
      }
      comm_world.abort(0); //      MPI_Abort(comm,0);
    }
  }

  /*
   * Every process gets its own data and sends it to process zero.
   * Exercise: why are we using an Isend here?
   */
  int mydata = procno+1;
  //  MPI_Request send_request;
  auto send_request = comm_world.isend(mydata,0);
  
  /*
   * Now process zero receives all the contribution using the subarray type.
   * (You could use a gather here, but only because it is just one element)
   */
  if (procno==0) {
    vector<int> cubedata(procs_per_side*procs_per_side*procs_per_side);
    int sides[3], sub_sizes[3], startpoint[3];
    for (int proci=0; proci<procs_per_side; proci++) {
      for (int procj=0; procj<procs_per_side; procj++) {
	for (int prock=0; prock<procs_per_side; prock++) {
	  /*
	   * Now receive the data, using a subarray type to
	   * put it in the right location of the cubedata buffer.
	   */
/**** your code here ****/
	  //MPI_Datatype insert_type;
	  mpl::subarray_layout<int> insert_type
	    ( { 
	      { procs_per_side,1,proci },
	      { procs_per_side,1,procj },
	      { procs_per_side,1,prock }
	    } );
	  /*
	   * Now do the receive call
	   */
	  int sender = ( proci*procs_per_side + procj )*procs_per_side + prock;
	  comm_world.recv
	    (
/**** your code here ****/
	     sender);
	}
      }
    }

    proctext << "Received cube:";
    int error{0};
    for (int p=0; p<nprocs; p++) {
      if (cubedata[p]!=p+1)
	error++;
      proctext << " " << cubedata[p];
    }
    proctext << "\n";
    if (error==0)
      proctext << "Finished. Cube correctly gathered.";
    else
      proctext << "There were errors";
    proctext << endl;
    cerr << proctext.str(); 
  }
  send_request.wait();
  
  return 0;
}
